crates/sui-package-resolver/src/store.rs

```diff
@@ -0,0 +1,225 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::BTreeMap;
+use std::path::Path;
+use std::sync::Arc;
+
+use crate::error::Error;
+use crate::Result;
+use crate::{Module, Package};
+use async_trait::async_trait;
+use diesel::{ExpressionMethods, OptionalExtension, QueryDsl, RunQueryDsl};
+use move_binary_format::{access::ModuleAccess, errors::Location, CompiledModule};
+use move_core_types::account_address::AccountAddress;
+use sui_indexer::{indexer_reader::IndexerReader, schema_v2::objects};
+use sui_rest_api::Client;
+use sui_types::base_types::ObjectID;
+use sui_types::{base_types::SequenceNumber, move_package::TypeOrigin, object::Object};
+use typed_store::rocks::{DBMap, MetricConf};
+use typed_store::traits::TableSummary;
+use typed_store::traits::TypedStoreDebug;
+use typed_store::Map;
+use typed_store_derive::DBMapUtils;
+
+/// Interface to abstract over access to a store of live packages.  Used to override the default
+/// store during testing.
+#[async_trait]
+pub trait PackageStore {
+    /// Latest version of the object at `id`.
+    async fn version(&self, id: AccountAddress) -> Result<SequenceNumber>;
+
+    /// Read package contents. Fails if `id` is not an object, not a package, or is malformed in
+    /// some way.
+    async fn fetch(&self, id: AccountAddress) -> Result<Package>;
+
+    /// Store package `object` in the underlying store
+    async fn update(&self, object: &Object) -> Result<()>;
+}
+
+pub struct DbPackageStore(pub(crate) IndexerReader);
```

Comment:
In order to have a better separation of dependencies I think that we shouldn't have any implementations of "PackageStore" (except for maybe a purely in-memory one) live in this crate. Instead we should export the trait and then let downstream consumers build and implement their own package caches using the dependencies that they want to bring in.

As it stands in this PR right now, in order to use this package resolver you need to depend on rocksdb and postgres despite the analytics pipeline not needing to depend on postgres and the graphql servers not needing to depend on rocksdb.

Alternatively we can leverage feature flags to bring those implementations and dependencies...

I mean today we don't really worry much about these things so we could punt on this but eventually we as a project need to do a bit better with how we're structuring these things. 

@amnn What are your thoughts on this?

Reply:
> Alternatively we can leverage feature flags

Please, no 😭 .

Reply:
The setup that seems to work best is to:
- Have a `core` base crate which defines the interface and maybe one simple, no dependency, implementation
- All implementations exist in their own small crate or live in the crate of where they're consumed/used
- You can have a mega "sdk" like crate that uses feature flags to enable the implementations that the user wants by optionally enabling the dependencies where those implementations exist, and rexporting them. this gives you a single crate that users can depend on.

That last one is not very good for our own usecases internally but could be interesting for external user. My recommendation is we do the first two bullets and ignore the third for now. Which seems to be inline with your preferences too.